libass/ass_drawing.c

```diff
@@ -43,93 +43,148 @@ static bool token_check_values(ASS_DrawingToken *token, int i, ASS_TokenType typ
+    size_t coords = 0;
 
     while (p && *p) {
-        int got_coord = 0;
-        if (*p == 'c' && spline_start) {
-            // Close b-splines: add the first three points of the b-spline
-            // back to the end
-            if (token_check_values(spline_start->next, 2, TOKEN_B_SPLINE)) {
-                for (int i = 0; i < 3; i++) {
-                    tail->next = calloc(1, sizeof(ASS_DrawingToken));
-                    tail->next->prev = tail;
-                    tail = tail->next;
-                    tail->type = TOKEN_B_SPLINE;
-                    tail->point = spline_start->point;
-                    spline_start = spline_start->next;
-                }
-                spline_start = NULL;
-            }
-        } else if (!is_set && mystrtod(&p, &val)) {
-            point.x = double_to_d6(val);
-            is_set = 1;
-            got_coord = 1;
-            p--;
-        } else if (is_set == 1 && mystrtod(&p, &val)) {
-            point.y = double_to_d6(val);
-            is_set = 2;
-            got_coord = 1;
-            p--;
-        } else if (*p == 'm')
-            type = TOKEN_MOVE;
-        else if (*p == 'n')
-            type = TOKEN_MOVE_NC;
-        else if (*p == 'l')
-            type = TOKEN_LINE;
-        else if (*p == 'b')
-            type = TOKEN_CUBIC_BEZIER;
-        else if (*p == 'q')
-            type = TOKEN_CONIC_BEZIER;
-        else if (*p == 's')
-            type = TOKEN_B_SPLINE;
-        // We're simply ignoring TOKEN_EXTEND_B_SPLINE here.
-        // This is not harmful at all, since it can be ommitted with
-        // similar result (the spline is extended anyway).
-
-        // Ignore the odd extra value, it makes no sense.
-        if (!got_coord)
-            is_set = 0;
-
-        if (type != TOKEN_INVALID && is_set == 2) {
-            if (root) {
-                tail->next = calloc(1, sizeof(ASS_DrawingToken));
-                tail->next->prev = tail;
-                tail = tail->next;
-            } else {
-                /* VSFilter compat:
-                 * In guliverkli(2) VSFilter all drawings
-                 * whose first valid command isn't m are rejected.
-                 * xy-VSF and MPC-HC ISR this was (possibly inadvertenly) later relaxed,
-                 * such that all valid commands but n are ignored if there was no m yet.
-                 */
-                if (type == TOKEN_MOVE_NC) {
-                    return NULL;
-                } else if (type != TOKEN_MOVE) {
-                    p++;
+        char cmd = *p;
+        p++;
+        /* VSFilter compat:
+         * In guliverkli(2) VSFilter all drawings
+         * whose first valid command isn't m are rejected.
+         * xy-VSF and MPC-HC ISR this was (possibly inadvertenly) later relaxed,
+         * such that all valid commands but n are ignored if there was no m yet.
+         * 'p' is further restricted and ignored unless there arealready >= 3 nodes.
```

Comment:
*already

Reply:
also, clarify if the p restriction is new in xy/ISR or not

Reply:
thx, fixed the typo and added *“in newer VSFs, but guliverkli2 just crashed on 'p' with  1 <= n < 3 preceding nodes”* to the `p` description

Reply:
Is there a reason to comment on the crashing for `p` but not the other commands? Older VSFilter crashed on each of these commands if there weren’t enough preceding points; that’s why the guards were added.

Which makes me realize: our handling of “required first `m`” may be still wrong in master and after this PR in its current form. VSFilter actually requires an `m` command _letter_ to be seen at a point where no valid commands have been read yet, but the `m` command itself doesn’t have to be valid (to have valid coordinates)! This is probably most of the reason for those crashes they had to fix, lol.

Reply:
Specifically, the `root` can actually be an `n` command:

```
m n 0 0 l 100 0 100 100 0 100 0 0
```

shows a square in VSFilter identical to this:

```
m 0 0 l 100 0 100 100 0 100
```

Reply:
> Is there a reason to comment on the crashing for p but not the other commands? Older VSFilter crashed on each of these commands if there weren’t enough preceding points

Your second paragraph (good catch!) contains the reason why. If, as I mistakenly thought before, the drawing must start with a *valid* `m`, all those other crashes wouldn’t have been possible, since there already is one `m` and all other commands only need `>= 1` preceding node.

Adjusted the comment and implemented the “invalid `m` before `n` case”, thanks!